sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Reflect.scala

```diff
@@ -0,0 +1,174 @@
+ *                 and the remaining are input arguments to the Java method.
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(class,method[,arg1[,arg2..]]) calls method with reflection",
+  extended = "> SELECT _FUNC_('java.util.UUID', 'randomUUID');\n c33fb387-8500-4bfa-81d2-6e0e3e930df2")
+// scalastyle:on line.size.limit
+case class Reflect(children: Seq[Expression])
+  extends Expression with CodegenFallback {
+
+  override def prettyName: String = "reflect"
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size < 2) {
+      TypeCheckFailure("requires at least two arguments")
+    } else if (!children.take(2).forall(e => e.dataType == StringType && e.foldable)) {
+      // The first two arguments must be string type.
+      TypeCheckFailure("first two arguments should be string literals")
+    } else if (!classExists) {
+      TypeCheckFailure(s"class $className not found")
+    } else if (method == null) {
+      TypeCheckFailure(s"cannot find a method that matches the argument types in $className")
+    } else {
+      TypeCheckSuccess
+    }
+  }
+
+  override def deterministic: Boolean = false
+  override def nullable: Boolean = true
+  override val dataType: DataType = StringType
+
+  override def eval(input: InternalRow): Any = {
+    var i = 0
+    while (i < argExprs.length) {
+      buffer(i) = argExprs(i).eval(input).asInstanceOf[Object]
+      // Convert if necessary. Based on the types defined in typeMapping, string is the only
+      // type that needs conversion. If we support timestamps, dates, decimals, arrays, or maps
+      // in the future, proper conversion needs to happen here too.
+      if (buffer(i).isInstanceOf[UTF8String]) {
+        buffer(i) = buffer(i).toString
+      }
+      i += 1
+    }
+    UTF8String.fromString(String.valueOf(method.invoke(obj, buffer : _*)))
+  }
+
+  @transient private lazy val argExprs: Array[Expression] = children.drop(2).toArray
+
+  /** Name of the class -- this has to be called after we verify children has at least two exprs. */
+  @transient private lazy val className = children(0).eval().asInstanceOf[UTF8String].toString
+
+  /** True if the class exists and can be loaded. */
+  @transient private lazy val classExists = Reflect.classExists(className)
+
+  /** The reflection method. */
+  @transient lazy val method: Method = {
+    val methodName = children(1).eval(null).asInstanceOf[UTF8String].toString
+    Reflect.findMethod(className, methodName, argExprs.map(_.dataType)).orNull
+  }
+
+  /** If the class has a no-arg ctor, instantiate the object. Otherwise, obj is null. */
+  @transient private lazy val obj: Object =
+    Reflect.instantiate(className).orNull.asInstanceOf[Object]
+
+  /** A temporary buffer used to hold intermediate results returned by children. */
+  @transient private lazy val buffer = new Array[Object](argExprs.length)
```

Comment:
It doesn't matter how many arguments there are, but the array is always created. This is what Hive does too. I don't buy the premature optimization argument, since this is pretty simple to do.